lightning/src/ln/channelmanager.rs

```diff
@@ -2282,6 +2297,12 @@ impl<ChanSigner: ChannelKeys, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref>
 	/// ChannelMonitorUpdateErr::TemporaryFailures is fine. The highest_applied_update_id field
 	/// exists largely only to prevent races between this and concurrent update_monitor calls.
 	///
+	/// In some cases, this may generate a monitor update, resulting in a call to the
+	/// `chain::Watch`'s `update_channel` method for the same channel monitor which is being
+	/// notified of a successful update here. Because of this, please be very careful with
+	/// reentrancy bugs! It is incredibly easy to write an implementation of `update_channel` which
+	/// will take a lock which is also held when calling this method.
+	///
```

Comment:
Only in the sense that we're updating precisely the ChannelMonitor (via `update_channel`) that we're being notified has been successfully updated. I'm not quite sure how to better capture this in the docs here, but its rather easy (as we do in tests), to hit this case (but only rarely because you only ever get the callbacks rarely).
```
struct TestChainMonitor {
    monitor_state: Mutex<HashMap<OutPoint, StateInformation>>,
}
impl TestChainMonitor {
    fn finish_async_update() {
        let monitor_state = self.monitor_state.lock().unwrap();
        ...
        channel_manager.channel_monitor_updated(...);
    }
}
impl chain::Watch for TestChainMonitor {
    fn update_channel(&self, funding_txo: OutPoint, update: channelmonitor::ChannelMonitorUpdate) -> Result<(), channelmonitor::ChannelMonitorUpdateErr> {
        let this_monitor_state = self.monitor_state.lock().unwrap();
        ...
        start_async_update(update);
        Err(TemporaryFailure)
    }
    ....
}
```